core/src/main/java/bio/terra/pearl/core/service/address/AddressValidationClientStub.java

```diff
@@ -0,0 +1,88 @@
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.UUID;
+
+@Component
+@Slf4j
+public class AddressValidationClientStub implements AddressValidationClient {
+
+    private static final String BAD_ADDRESS_INDICATOR = "BAD";
+    private static final String IMPROVABLE_ADDRESS_INDICATOR = "IMPROVABLE";
+    private static final String ERROR_INDICATOR = "ERROR";
+
+    /**
+     * Stubbed client allows you to test various possible states of address validation.
+     * <br>
+     * State 1: Invalid address
+     * - Put the word "BAD" in street1, e.g. 123 Bad St
+     * - All the tokens in street1 will be considered unresolved
+     * - Example missing component data will be returned
+     * - Real-world responses are not guaranteed have missingComponents or unresolvedTokens
+     * <br>
+     * State 2: Improvable address
+     * - Put the word "IMPROVABLE" in street1, e.g. 415 IMPROVABLE St
+     * - Returns valid = true with a suggested address of 415 Main St
+     * <br>
+     * State 3: Server side error
+     * - Put the word "ERROR" in street1, e.g. 500 ERROR Lane
+     * - Throws an AddressValidationException
+     * - HTTP Status code defaults to 500, but will use a valid 400-599 status code if present in postalCode
+     * <br>
+     * State 4: Valid address
+     * - All other addresses will default to valid
+     */
+    @Override
+    public AddressValidationResultDto validate(UUID sessionId, MailingAddress address) throws AddressValidationException {
+
+        if (address.getStreet1().contains(BAD_ADDRESS_INDICATOR)) {
+            return AddressValidationResultDto
+                    .builder()
+                    .valid(false)
+                    .missingComponents(List.of("street", "country", "postal_code")) // junk example data
+                    .unresolvedTokens(List.of(StringUtils.split(address.getStreet1())))
+                    .sessionId(sessionId)
+                    .build();
+        } else if (address.getStreet1().contains(IMPROVABLE_ADDRESS_INDICATOR)) {
+            return AddressValidationResultDto
+                    .builder()
+                    .valid(true)
+                    .suggestedAddress(
+                            MailingAddress
+                                    .builder()
+                                    .street1("415 Main St")
+                                    .city("Cambridge")
+                                    .state("MA")
+                                    .country("USA")
+                                    .postalCode("02142")
+                                    .build())
+                    .sessionId(sessionId)
+                    .build();
+        } else if (address.getStreet1().contains(ERROR_INDICATOR)) {
+            int errorCode = 500;
+            try {
+                errorCode = Integer.parseInt(address.getPostalCode());
```

Comment:
I know this is a stub, but I don't understand this line, how/why is the error code relevant to the postal code? In case of server error on Juniper side (like the case in the catch) do you want to return 404 or 500?

Reply:
Oh also something that may be useful, Postal codes in Canada are a six-character string consisting of three letters and three digits. For example, the postal code for the CN Tower in Toronto is M5V 3L9.

Reply:
It allows you to customize the error returned so you can test different scenarios. E.g.

```
123 ERROR St
Boston, MA 404
```
returns a 404 error

Reply:
Gotcha